File: 1172-dinner-plate-stacks/1172-dinner-plate-stacks.js
/**
 * @param {number} capacity
 */
var DinnerPlates = function(capacity) {
  this.capacity = capacity;
  this.stacks = [];
  this.pq = new MinPriorityQueue();
};

/** 
 * @param {number} val
 * @return {void}
 */
DinnerPlates.prototype.push = function(val) {
  if (!this.pq.isEmpty() && this.pq.front().element >= this.stacks.length) this.pq.clear();
  if (!this.pq.isEmpty()) {
    const top = this.pq.front().element;
    this.stacks[top].push(val);
    
    if (this.stacks[top].length === this.capacity) this.pq.dequeue();
  }
  else {
    this.stacks.push([val]);
    if (this.capacity > 1) this.pq.enqueue(this.stacks.length - 1);
  }
};

/**
 * @return {number}
 */
DinnerPlates.prototype.pop = function() {
  return this.popAtStack(this.stacks.length - 1);
};

/** 
 * @param {number} index
 * @return {number}
 */
DinnerPlates.prototype.popAtStack = function(index) {
  if (index < 0 || index >= this.stacks.length || !this.stacks[index].length)
    return -1;
  
  if (this.stacks[index].length === this.capacity) this.pq.enqueue(index);
  
  const val = this.stacks[index].pop();
  
  while (this.stacks.length && !this.stacks.at(-1).length) this.stacks.pop();
  
  return val;
};

/** 
 * Your DinnerPlates object will be instantiated and called as such:
 * var obj = new DinnerPlates(capacity)
 * obj.push(val)
 * var param_2 = obj.pop()
 * var param_3 = obj.popAtStack(index)
 */

File: 1172-dinner-plate-stacks/1172-dinner-plate-stacks.py
class DinnerPlates:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.stacks = []
        self.h = []

    def push(self, val: int) -> None:
        if self.h and self.h[0] >= len(self.stacks):
            self.h = []
        if self.h:
            self.stacks[self.h[0]].append(val)
            if len(self.stacks[self.h[0]]) == self.capacity:
                heappop(self.h)
        else:
            self.stacks.append([val])
            if self.capacity > 1:
                heappush(self.h, len(self.stacks) - 1)

    def pop(self) -> int:
        return self.popAtStack(len(self.stacks) - 1)

    def popAtStack(self, index: int) -> int:
        if index < 0 or index >= len(self.stacks) or not self.stacks[index]:
            return -1
        if len(self.stacks[index]) == self.capacity:
            heappush(self.h, index)
        val = self.stacks[index].pop()
        while self.stacks and not self.stacks[-1]:
            self.stacks.pop()
        return val
        


# Your DinnerPlates object will be instantiated and called as such:
# obj = DinnerPlates(capacity)
# obj.push(val)
# param_2 = obj.pop()
# param_3 = obj.popAtStack(index)